File: examples/python/templates/template.py
import sys

class Token:
    ERROR = "error"
    EMPTY = "empty"
    {%- for scanner in scanners %}
    {{ scanner.token|upper }} = "{{ scanner.token|lower }}"
    {%- endfor %}

class InputStack:
    def __init__(self, file=sys.stdin):
        self.stack = []
        self.file = file
        self.done = False

    def get(self):
        if self.stack:
            return self.stack.pop()
        if self.done:
            return ""
        a = self.file.read(1)
        if not a:
            self.done = True
        return a

    def unget(self, a):
        if not a:
            raise Exception
        self.stack.append(a)

class BaseRecognizer:
    def __init__(self, token=Token.EMPTY, accept=False, error=-1):
        self.token = token
        self.accept = accept
        self.error = error

    def next(self, q, a):
        raise NotImplementedError

    def match(self, io=InputStack()):
        checkpoint = [0]
        lexeme = []
        while checkpoint[-1] != self.error:
            a = io.get()
            if not a:
                break
            status, r = self.next(checkpoint[-1], ord(a))
            if status == 1:
                self.accept = True
                checkpoint.clear()
            checkpoint.append(r)
            lexeme.append(a)
        for _ in range(len(checkpoint), 1, -1):
            io.unget(lexeme.pop())
        return self.accept, "".join(lexeme)

{% for scanner in scanners %}
{% include "scanner.py" %}
{% endfor %}
def match_first(*recs, io=InputStack()):
    for T in recs:
        r = T()
        ok, s = r.match(io)
        if ok:
            return r.token, s
    return Token.EMPTY, ""

def match_longest(*recs, io=InputStack()):
    token = Token.ERROR
    lexeme = ""
    for T in recs:
        r = T()
        ok, s = r.match(io)
        if ok and len(s) > len(lexeme):
            token = r.token
            lexeme = s
        for a in s[-1::-1]:
            io.unget(a)
    for _ in lexeme:
        io.get()
    if token == Token.ERROR and io.done:
        token = Token.EMPTY
    return token, lexeme

SCANNERS = {
{%- for scanner in scanners %}
    "{{ scanner.token }}": {{ scanner.token|title }},
{%- endfor %}
}

File: examples/poc/cpp2/templates/template.cpp
#pragma once
#include <cstdint>
#include <iostream>
#include <memory>
#include <string>
#include <utility>
#include <vector>

#define ALL_RECOGNIZERS \
    {%- for scanner in scanners %}
    {{ scanner.token|title }}{% if loop.index != loop.length %}, \{% endif -%}
    {% endfor %}

enum class Token {
    EMPTY = 0,
    {%- for scanner in scanners %}
    {{ scanner.token|upper }},
    {%- endfor %}
};

std::ostream& operator<<(std::ostream& out, Token token)
{
    switch (token) {
    case Token::EMPTY:
        return out << "empty";
    {%- for scanner in scanners %}
    case Token::{{ scanner.token|upper }}:
        return out << "{{ scanner.token }}";
    {%- endfor %}
    default:
        return out;
    }
}

struct BaseRecognizer {
    Token token;

    BaseRecognizer(
        Token token = Token::EMPTY,
        bool accept = false, // should be true if 0 is an accept state
        int error = -1)
        : token(token)
        , accept(accept)
        , error(error)
    {}

    //virtual ~BaseRecognizer() {}    // TODO is this needed?
    virtual std::pair<int, int> next(int q, uint32_t a) const = 0;

    std::pair<bool, std::string> match(std::istream& in = std::cin)
    {
        std::vector<int> checkpoint = {0};
        std::vector<uint32_t> lexeme;
        uint32_t a = -1;
        uint32_t eof = std::char_traits<char>::eof();
        while (checkpoint.back() != error && (a = in.get()) != eof) {
            auto [status, r] = next(checkpoint.back(), a);
            if (status == 1) {
                accept = true;
                checkpoint.clear();
            }
            checkpoint.push_back(r);
            lexeme.push_back(a);
        }
        for (auto i = checkpoint.size(); i > 1; --i) {
            // rollback to most recent accept state, if any
            // (if accept is true, checkpoint[0] is the most recent accept state)
            in.unget();
            lexeme.pop_back();
        }
        return { accept, std::string(lexeme.begin(), lexeme.end()) };
    }

protected:
    bool accept;
    int const error;
};
{% for scanner in scanners %}
{% include "scanner.cpp" %}
{% endfor %}
template <class... Recognizers>
std::pair<Token, std::string> match_first(std::istream& in = std::cin)
{
    std::vector<std::unique_ptr<BaseRecognizer>> recs = {
        std::make_unique<Recognizers>() ...
    };
    for (auto r: recs) {
        auto [ok, s] = r->match(in);
        if (ok) {
            return {r->token, s};
        }
    }
    return {Token::EMPTY, ""};
}

template <class... Recognizers>
std::pair<Token, std::string> match_longest(std::istream& in = std::cin)
{
    std::vector<std::unique_ptr<BaseRecognizer>> recs = {
        std::make_unique<Recognizers>() ...
    };
    Token token = Token::EMPTY;
    std::string lexeme;
    for (auto r: recs) {
        auto [ok, s] = r->match(in);
        if (ok && s.size() > lexeme.size()) {
            token = r->token;
            lexeme = s;
        }
    }
    return {token, lexeme};
}
